### PR TITLE
Some test-related fixed and a test runner for installed tests

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Comment by yuwata** [comment]
Note about the message of the commit renaming `test-udev`:
> We don't install this test executable

We install the test... For fedora, systemd-tests.rpm provides `test-udev` under /usr/lib/systemd/tests/manual. Though, the binary is almost useless, as `udev-test.pl` is not installed...

File: src/test/meson.build [hunk 1]
```diff
@@ -620,7 +620,7 @@ tests += [
          [libshared],
          []],
 
-        [['src/test/test-udev.c'],
+        [['src/test/udev-runner.c'],
```
Comment (yuwata) [inline src/test/meson.build]: I do not like this change. I think all tests should be prefixed with 'test-'. As already commented, `test-udev` will be installed under `manual` directory, at least on fedora. So, it is safely excluded from the list in the driver provided by the later commit.

File: src/shared/seccomp-util.c [hunk 2]
```diff
@@ -878,25 +876,26 @@ int seccomp_add_syscall_filter_item(scmp_filter_ctx *seccomp, const char *name,
...
+                if (id < 0)
+                        /* No such system call exists on this architecture. */
+                        return 0;
+
                 r = seccomp_rule_add_exact(seccomp, action, id, 0);
                 if (r < 0)
-                        /* If the system call is not known on this architecture, then that's fine, let's ignore it */
-                        log_debug_errno(r, "Failed to add rule for system call %s() / %d, ignoring: %m", name, id);
+                        return log_debug_errno(r, "Failed to add rule for system call %s() / %d, ignoring: %m", name, id);
+                return 0;
```
Comment (yuwata) [inline src/shared/seccomp-util.c]: This seems to change the behavior.

File: src/shared/seccomp-util.c [hunk 3]
```diff
@@ -977,18 +976,22 @@ int seccomp_load_syscall_filter_set_raw(uint32_t default_action, Hashmap* set, u
 
                 HASHMAP_FOREACH_KEY(val, id, set, i) {
                         uint32_t a = action;
-                        int e = PTR_TO_INT(val);
+                        int syscall = PTR_TO_INT(id) - 1;
```
Comment (yuwata) [inline src/shared/seccomp-util.c]: I think this should be renamed to e.g. `syscall_id` or something, as this 'conflicts' to `syscall()`.

**Comment by keszybz** [comment]
I update the commit message to describe the changed behaviour (this was on purpose, the commit description was wrong), and renamed the variable and pushed another commit on top to run unsafe tests optionally.

File: src/shared/seccomp-util.c (hunk 2 above)
Comment (yuwata) [reply src/shared/seccomp-util.c]: Sorry, what I want to say in the previous comment is that I guess that this behavior change causes the previous CI failures. But I am not sure. Let's wait the currently running CIs.

BTW, why do you propagate the error here, but not in `seccomp_load_syscall_filter_set_raw()`? It seems not consistent for me...

Comment (keszybz) [reply src/shared/seccomp-util.c]: We were ignoring all errors, and that means that "legitimate" errors could be hidden in between the trivial errors about unimplemented syscalls. I'll wait for CI this time.

> BTW, why do you propagate the error here, but not in `seccomp_load_syscall_filter_set_raw`()? It seems not consistent for me...

Yep, I'll change `seccomp_load_syscall_filter_set_raw` too.

**Comment by yuwata** [comment]
Hmm, many tests are failed. E.g.,
```
Spawning container nspawn-root on /var/tmp/systemd-test.m76vlQ/nspawn-root.
Press ^] three times within 1s to kill container.
Failed to add syscall filter item @file-system: Numerical argument out of domain
E: nspawn failed with exit code 1
ls: cannot access '/var/tmp/systemd-test.m76vlQ/journal/*/*.journal': No such file or directory
TEST RUN: Tmpfiles related tests [FAILED]
make: *** [run] Error 1
Makefile:4: recipe for target 'run' failed
make: Leaving directory '/tmp/autopkgtest.Fvbsym/build.Z13/debian/test/TEST-22-TMPFILES'
```

**Comment by keszybz** [comment]
I removed the seccomp part from this PR. I'll re-submit that separately.

File: test/run-unit-tests.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+tests = glob.glob('/usr/lib/systemd/tests/test-*')
+total = Total(total=len(tests))
+for test in tests:
+    name = os.path.basename(test)
+
+    ex = subprocess.run(test, stdin=subprocess.DEVNULL, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    if ex.returncode == 0:
+        print(f'{GREEN}PASS: {name}{RESET_ALL}')
+        total.good += 1
```
Comment (ssahani) [inline test/run-unit-tests.py:45]: Green/RED no no

Comment (ssahani) [reply test/run-unit-tests.py:45]: unittest it much better than this

Comment (ssahani) [reply test/run-unit-tests.py:45]: >- @ssahani's test runner [2]
  It uses the unittest library and the test suite was implented as a class, and
  doesn't implement any of the functions listed above.

every test function is generated dynamically if that is a test and  unit test is standard.

**Comment by keszybz** [comment]
> Why do you want invent yourself when python halve itself own test framework. This makes no sense

This is all explained in the commit message. Is anything unclear?

> Why python why not c 

I think you would have to explain why I should use C when a a scripting language is enough.

**Comment by ssahani** [comment]
> > Why do you want invent yourself when python halve itself own test framework. This makes no sense
> 
> This is all explained in the commit message. Is anything unclear?
See my ans below
> 
> > Why python why not c
> 
> I think you would have to explain why I should use C when a a scripting language is enough.
I asked you. The point is unittest is much better

**Comment by mbiebl** [comment]
> Note about the message of the commit renaming `test-udev`:
> 
> > We don't install this test executable
> 
> We install the test... For fedora, systemd-tests.rpm provides `test-udev` under /usr/lib/systemd/tests/manual. Though, the binary is almost useless, as `udev-test.pl` is not installed...

Fwiw, the binary *is* used on Debian/Ubuntu, see
https://salsa.debian.org/systemd-team/systemd/blob/master/debian/tests/udev

**Comment by keszybz** [comment]
There have been various general questions, I'll try to answer them in one go.

Our unit test consists of a series of binaries that follow the "meson test" protocol (for example 77 return value has special meaning). This protocol is not directly supported by any tool except for meson test and autotools test runner, but they are designed for running from the build tree, and we want to run installed tests. This means that we need a special runner that understand the protocol to find and run the tests. The first question was whether to write it in C or shell or python. A high-level language is easier and faster, and there is no particular reason to use C. I prefer python because I think the code is easier to read, and adding functionality like an argument parser is very easy.

The second question whether to use unittest.  We can use fully custom code to do this (like this PR proposes and like the Ubuntu runner does), or we could use a framework. If we use a framework, we still need to do the discovery of tests and protocol handling, because our tests are not python functions. This means that we wouldn't be using most of the functionality of the test framework. I think it doesn't make sense to pull in a framework at this point.

Re `test-udev`: I already removed the commit that renamed it.

**Comment by yuwata** [comment]
> I prefer python because I think the code is easier to read, and adding functionality like an argument parser is very easy.

I agree with these points. However, it is hard to use the driver from our test suite TEST-XX-*. I thought I will use this in #10136, but copying python to container image seems hard. Of course, fundamentally it is possible, but bash script is easier to use.

**Comment by poettering** [comment]
Where do we stand on this?

I am fine with python for this, but then again, shell is probably good enough too to spawn a few tests...

@mbiebl you package the installed tests iirc. What's your take on this? do you mind if this depends on python?

this lgtm, but I'd like the debian maintainers input on this (or in fact any distro's input which packages the tests, anyone else?)

**Comment by poettering** [comment]
ok, let's merge this then. I figure if other downstreams prefer this to be a shell script they could send a follow-up PR, and we can reconsider.

File: test/run-unit-tests.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python3
+
+import argparse
+import dataclasses
```
Comment (mbiebl) [inline test/run-unit-tests.py:4]: The usage of dataclasses means it requires at least python 3.7, right?

**Comment by mbiebl** [comment]
@keszybz would it be possible to add a ```--verbose``` to the test-runnter to dump the output of the tests to stdout/stderr? This is really helpful when you run the unit tests non-interactively on your CI machinery.

Or even better: just dump the output of the failing tests?

**Comment by mbiebl** [comment]
Oh, nvm, Should have read the commit message more carefully. It already dumps the output on failure...

File: test/run-unit-tests.py (hunk 5 above)
Comment (keszybz) [reply test/run-unit-tests.py:4]: I guess. It you need to support older version, patches are welcome ;)

Comment (mbiebl) [reply test/run-unit-tests.py:4]: In Debian, 3.6 is the current default for python3, so I'll just keep our own test-runner for the time being.

**Pull request merged**